ocaml/testsuite/tests/typing-zero-alloc/signatures.ml

```diff
@@ -0,0 +1,821 @@
+(* TEST
```

Comment:
Other maybe interesting test cases:
  * `module type of` a structure that uses zero_alloc? I'm prepared for the answer to be "the attributes are dropped in the inferred signature" but even that feels worth recording in a test.
  * A test with a functor where the zero-allocness of the impl depends on the zero-allocness of functions in the param's signature? I'm unsure if this is supported yet. Regardless another test file might be a better home than this one.

Reply:
Right, I meant to add `module type of` tests, and have now done so. It works well. The fact that you can't sneak `assume` or `assume never_returns_normally` into a signature with it motivated parts of the details of the current implementation.

As for the functor case: There is currently no way for the signatures to influence the zero alloc checking of implementations, beyond requiring that such a check might happen.  So I don't think we can do your example.  But I will add lots of tests like this in the next PR.